### PR TITLE
rustc: Add `Local` to the HIR map of parents

### LINKED ISSUE #43910: Nightly version ignores allow(unused)
The current nightly version (rustc 1.21.0-nightly (df511d554 2017-08-14)) warns about an unused variable, even if it's flagged with allow(). The stable version doesn't do so.
```
fn main() {
    #[allow(unused_variables)] let x = 12;
    let y = 11;
}
```

```
warning: unused variable: `x`
 --> src/main.rs:2:36
  |
2 |     #[allow(unused_variables)] let x = 12;
  |                                    ^
  |
  = note: #[warn(unused_variables)] on by default
```

https://play.rust-lang.org/?gist=9b81da266dd1f4383760de35d94f7ff6&version=nightly

### PR DESCRIPTION
**Pull request by alexcrichton**
When walking parents for lints we want to be sure to hit `let` statements which
can have attributes, so hook up these statements in the HIR map.

Closes #43910

### CODE REVIEW

File: src/test/run-pass/issue-43910.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2017cThe Rust Project Developers. See the COPYRIGHT
```
Comment (kennytm) [inline src/test/run-pass/issue-43910.rs]: Needs to remove the "c".

```
[00:03:51] tidy error: /checkout/src/test/run-pass/issue-43910.rs: incorrect license
```

**Comment by kennytm** [comment]
The UI test for `issue-40402-ref-hints` needs to be updated, but I'm not sure if this is intended (the old suggestion seems more idiomatic).

```diff
 error[E0507]: cannot move out of indexed content
   --> $DIR/issue-40402-1.rs:19:13
    |
 19 |     let e = f.v[0];
-   |             ^^^^^^
-   |             |
-   |             cannot move out of indexed content
-   |             help: consider using a reference instead: `&f.v[0]`
+   |         -   ^^^^^^ cannot move out of indexed content
+   |         |
+   |         hint: to prevent move, use `ref e` or `ref mut e`
 
 error: aborting due to previous error
```

**Comment by alexcrichton** [comment]
Fixed

**Comment by michaelwoerister** [comment]
@bors r+

Thanks, Alex!

**Comment by kennytm** [comment]
@bors retry 

`x86_64-gnu-aux` failed to clone `xsv`. Temporary GitHub issue, status report at:

* https://status.github.com/messages
* https://www.traviscistatus.com/incidents/tjfn9tf5vp83

<details>

```
[00:46:56] testing https://github.com/BurntSushi/xsv
[00:46:56] Initialized empty Git repository in /checkout/obj/build/ct/xsv/.git/
[00:46:56] fatal: Could not parse object 'a9a7163f2a2953cea426fee1216bec914fe2f56a'.
[00:46:56] From https://github.com/BurntSushi/xsv
[00:46:56]  * branch            master     -> FETCH_HEAD
[00:46:56] fatal: Could not parse object 'a9a7163f2a2953cea426fee1216bec914fe2f56a'.
[00:46:57] From https://github.com/BurntSushi/xsv
[00:46:57]  * branch            master     -> FETCH_HEAD
[00:46:57] fatal: Could not parse object 'a9a7163f2a2953cea426fee1216bec914fe2f56a'.
[00:50:31] fatal: unable to access 'https://github.com/BurntSushi/xsv/': Failed to connect to github.com port 443: Connection timed out
[00:50:31] thread 'main' panicked at 'assertion failed: status.success()', /checkout/src/tools/cargotest/main.rs:106:12
[00:50:31] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:50:31] 
[00:50:31] 
[00:50:31] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/cargotest" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build/ct"
[00:50:31] expected success, got: exit code: 101
[00:50:31] 
[00:50:31] 
[00:50:31] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/cargotest src/tools/cargo src/tools/rls src/test/pretty src/test/run-pass/pretty src/test/run-fail/pretty src/test/run-pass-valgrind/pretty src/test/run-pass-fulldeps/pretty src/test/run-fail-fulldeps/pretty
[00:50:31] Build completed unsuccessfully in 0:03:39
[00:50:31] make: *** [check-aux] Error 1
[00:50:31] Makefile:56: recipe for target 'check-aux' failed
```

</details>

**Comment by kennytm** [comment]
@bors retry #40672 

* Build: https://travis-ci.org/rust-lang/rust/builds/266825511
* Job `check i686-apple-darwin` not started: https://travis-ci.org/rust-lang/rust/jobs/266825517

**Pull request merged**